activerecord/lib/active_record/connection_handling.rb

```diff
@@ -96,7 +96,7 @@ def connection
     # Return the specification name from the current class or its parent.
     def connection_specification_name
       if !defined?(@connection_specification_name) || @connection_specification_name.nil?
-        @connection_specification_name = self == Base ? "primary" : superclass.connection_specification_name
+        return self == Base ? "primary" : superclass.connection_specification_name
       end
       @connection_specification_name
```

Comment:
Could take the opportunity to simplify this code:

```ruby
if defined?(@connection_specification_name)
  @connection_specification_name
else
  self == Base ? "primary" : superclass.connection_specification_name
end
```

By the way, maybe the ivar assignment was what screwed up a plain `super`.